packages/core/src/logger/logger.ts

```diff
@@ -1,144 +1,196 @@
+    }
+
+    return this.asyncStorage.run(newContext, mainCallback);
+  }
+
+  public trace(data: Record<string, any>): void;
+  public trace(message: string, data?: Record<string, any>): void;
+  public trace(message: any, data?: Record<string, any>): void {
+    this.log(LogLevel.TRACE, message, data);
+  }
+
+  public debug(data: Record<string, any> | Error): void;
+  public debug(message: string, data?: Record<string, any> | Error): void;
+  public debug(message: any, data?: Record<string, any> | Error): void {
+    this.log(LogLevel.DEBUG, message, data);
+  }
+
+  public info(data: Record<string, any> | Error): void;
+  public info(message: string, data?: Record<string, any> | Error): void;
+  public info(message: any, data?: Record<string, any> | Error): void {
+    this.log(LogLevel.INFO, message, data);
+  }
+
+  public warning(data: Record<string, any> | Error): void;
+  public warning(message: string, data?: Record<string, any> | Error): void;
+  public warning(message: any, data?: Record<string, any> | Error): void {
+    this.log(LogLevel.WARNING, message, data);
+  }
+
+  public error(data: Record<string, any> | Error): void;
+  public error(message: string, data?: Record<string, any> | Error): void;
+  public error(message: any, data?: Record<string, any> | Error): void {
+    this.log(LogLevel.ERROR, message, data);
+  }
+
+  public fatal(data: Record<string, any> | Error): void;
+  public fatal(message: string, data?: Record<string, any> | Error): void;
+  public fatal(message: any, data?: Record<string, any> | Error): void {
+    this.log(LogLevel.ERROR, message, data);
+  }
+
+  public log(level: LogLevel | number, data: Record<string, any> | Error): void;
+  public log(level: LogLevel | number, message: string, data?: Record<string, any> | Error): void;
+  public log(
+    level: LogLevel | number,
+    messageOrDataOrError: Record<string, any> | Error | string,
+    maybeDataOrError?: Record<string, any> | Error,
   ): void {
-    const tag = this.context.getStore();
-
-    if (tag) {
-      this.queueManager.log(
-        tag,
-        pluginOrLevel,
-        levelOrMessage,
-        messageOrParams,
-        paramsOrData,
-        maybeData,
-      );
-    } else if (this.globalLogger) {
-      this.globalLogger.log(
-        pluginOrLevel,
-        levelOrMessage,
-        messageOrParams,
-        paramsOrData,
-        maybeData,
-      );
+    let error: Error | undefined;
+    let data: Record<string, any> | undefined;
+    let message: string | undefined;
+    if (typeof messageOrDataOrError === 'string') {
+      message = messageOrDataOrError;
+      if (maybeDataOrError instanceof Error) {
+        error = maybeDataOrError;
+      } else {
+        data = maybeDataOrError;
+      }
+    } else if (messageOrDataOrError instanceof Error) {
+      error = messageOrDataOrError;
+    } else {
+      data = messageOrDataOrError;
     }
+
+    this.add({
+      level,
+      message,
+      data,
+      error,
+    });
```

Comment:
proč zmizely `params` / formátování message pomocí `vsprintf`? to chceš, to je hrozně užitečný!

tahle konstrukce resolvování těch argumentů mi teda nepřijde o moc přehlednější než to co jsem měl :-/

`error` má bejt oddělený? chápu že chcem umět zpracovat error, ale přišlo by mi lepší ho rovnou rozebrat na `message` a `data.stack`, ať už další kód nemusí řešit, že tam může bejt nějakej `error` - jak má třeba HTML handler zalogovat entry, kde je `message`, `data.stack` a ještě `error`? To jsou efektivně dvě kompletní entry..

Reply:
1. vsprintf -> Az v handleru, v loggeru nema co delat
2. error chces oddeleny pro pripady, kdy chces odlisne message a error.message (to treba pouzivam casto). Pripadny merge dohromady az v handleru, tady nema co delat
3. Jak to ma zalogovat? To je problem handleru :). Ale jsou to pripady, ktere se dejou a ktere realne v API umoznujeme

Reply:
1. vsprintf: noo... jako původně jsem to měl už v loggeru, protože argumenty jsou pole (i kdyby všechny byly skalární), a pole je objekt a může se změnit - tj. myslíš si, že zaloguješ zprávu X s parametry ABC, ale tam, kde se reálně ty parametry expandujou do tý zprávy, už to bude DEF. Zároveň jsi ty parametry vyhodil úplně - takže handler to udělat nemůže, protože ty parametry nedostane. Často při logování něčeho, kde chceš v message mít něco z nějaký proměnný, můžeš použít template literals - ale `vsprintf` ti dává nástroje, který se ručně v template literalu neřeší úplně hezky. Opět to byla nějaká optimalizace používání za cenu mírně zvýšený komplexity uvnitř.

2. a 3. To ale podle mě právě nechceš..
    - usecase 1: chceš zalogovat výjimku jak je, ale ještě si k ní přidat zprávu -> zaloguješ výjimku a zvlášť svoji zprávu (v pořadí který ti dává smysl)
    - usecase 2: chceš zalogovat nějakou custom message, která obsahuje zprávu výjimky -> použiješ `vsprintf` parametry nebo template literal (`'API call threw an error: %s', [e.message]`)

  Když to uděláš po tvým, jak se pak třeba bude handler rozhodovat, kterou message má použít pro spočítání nějakýho unikátního hashe tý entry (což HTML handler potřebuje kvůli identifikátorům dumpů)? Původní řešení to sice nechávalo na programátorovi, ale tím mu dávalo flexibilitu rozhodnout se, co v daný situaci dává smysl; tohle jen komplikuje spoustu kódu uvnitř.. co třeba když začneme ke všem log entries přidávat stack trace? bude v `data`? bude bokem jako další property log entry? která stack trace má v ten moment přednost? původní řešení error prostě normalizuje na message a trace a žádnej další kód už tyhle věci nemusí řešit..

Reply:
1 vyhral Dan, 2/3 ja